### PR TITLE
Add resilient Vietnam accommodation search with reconciliation

### LINKED ISSUE #1: Initial version
We need to make a tool to search vietnam accomodations (rooms, hotels, apartments, houses), it should be available as telegram bot similar to how we did in https://github.com/konard/documents-processing

We should use telegram user account + telegram bot account, or telegram bot only account if possible to use for monitoring in chats and public channels.

We should use services like booking.com, airbnb.com and so on, which are popular at vietnam, as well as group chats in Telegram in all languages.

We should collect at least 20 web services and at least 20 top telegram chats on the topic as data source, with ability to automatically search for top data sources, to update them in the future, with `/update_sources` command, by default we search top 20 of each, which are provably most popular.

For web search services we should use [browser-commander](https://github.com/link-foundation/browser-commander) to actually search via web UI.

We should collect up to 10 photos of each room/hotel/hostel/house and so on.

With ability to get the cheapest offer all recalculated in VND if other currencies are presented.

So telegram both with `/search --cheapest` should be able to answer with one best offer, or multiple of them with `/search --cheapest 10`.

All cached raw data, should be stored in links store and links notation as per https://github.com/link-assistant/formal-ai/blob/main/docs/associative-tech-stack.md (we also should use maximum other applicable components).

So we should do regular searches with presaving in cache up to 10 gb of data (photos included), we should be able to offload photos while preserving links to them to redownload them as needed.

So first execution of `/search --cheapest` should precache all latest data (see offers in telegram chats and posts only up to 2 months old), all converted to the same data format expressed as links, and after that we should use sorting by price (if needed currency conversion), we should check typical publicly available chats in telegram and make sure we support automatic parsing for messages with offers for as wide range of offers as possible.

### PR DESCRIPTION
**Pull request by konard**
Closes #1.

## Summary

- add a Telegram bot, CLI, and library services for browser-driven Vietnam accommodation searches
- keep three independent ranked cohorts: 20 web services, 20 nationwide Telegram communities, and 20 additional Nha Trang-focused Telegram communities
- include the requested `@Arenda_Vetnam`, `@Arenda_Nyachang_Zhilye`, and `@NaChangAp` sources and discover similar Nha Trang communities with English, Russian, and Vietnamese searches
- paginate public Telegram previews back to the strict two-month issue cutoff instead of reading only the currently rendered page
- normalize Russian, Vietnamese, and English listing data, including occupancy, room counts, area, floor, fees, dates, coordinates, reviews, amenities, contacts, and unfamiliar labeled fields
- reconcile listings transitively across canonical platform/official URLs, external IDs, source property IDs, and conservative location/structure/contact fingerprints
- preserve all identity aliases, source IDs and types, raw variants, identifiers, queries, contacts, attributes, photos, and per-source price observations when offers merge
- visit explicitly discovered accommodation websites directly, even when the parent marketplace card has no price, and give those observations dedicated `official-web` provenance
- detect official-site price increases and drops without mislabeling marketplace or Telegram prices as official
- compare prices in VND, return the cheapest offers, retain up to ten photos, and persist complete records with a shared 10 GiB Links Notation cache budget
- keep availability inquiries explicit and isolated through an optional `@mtcute/node` user session; ordinary searches never contact owners
- add the required minor changeset without editing the package version directly

## Reproduction and regression

Previously, records for the same accommodation remained separate when later observations arrived through a different source or URL alias. The store also had no durable distinction between general source-price observations and a price collected directly from the accommodation's official website.

The regression suites now prove that:

1. aliases merge transitively across collection runs while unrelated units sharing an agent remain separate;
2. both marketplace and official URLs remain valid identities, including when a later marketplace-only observation omits the official URL;
3. source provenance and raw variants survive reconciliation;
4. only direct `official-web` observations create official price-change events;
5. an explicitly marked official link is followed even when its marketplace card is unpriced;
6. arbitrary unlabeled URLs are not trusted as official-property websites;
7. unsafe literal local/private navigation targets are rejected;
8. adversarial official-URL punctuation is processed in bounded time without polynomial regex backtracking.

The minimum focused reproductions are in `tests/offer-reconciliation.test.js`, `tests/browser-dom-extraction.test.js`, and `tests/nha-trang-expansion.test.js`. Runtime and malformed-input boundaries are covered by `tests/core-edge-cases.test.js` and `tests/runtime-boundaries.test.js`.

## Usage

```bash
npm install
npx playwright install chromium
export TELEGRAM_BOT_TOKEN=your-bot-token
npm exec vietnam-accomodation-search -- bot
```

Example searches:

```text
/search --cheapest 10 Nha Trang
/search --cheapest 10 --filter bedrooms=2 --filter petsAllowed=true Nha Trang
/search --filter labeledFields.электричество=счётчику Нячанг
/update_sources
```

Availability inquiries additionally require an exported mtcute user session:

```bash
export TELEGRAM_API_ID=12345
export TELEGRAM_API_HASH=your-api-hash
export TELEGRAM_USER_SESSION=your-exported-session
node bin/vietnam-accomodation-search.js check-availability OFFER_ID @owner
```

The session string is equivalent to a password and must not be logged, committed, or shared. The same explicit action is available to the bot as `/check_availability OFFER_ID [@owner]`.

## Research and durable findings

- [`docs/COMPETITOR-RESEARCH.md`](https://github.com/konard/vietnam-accomodation-search/blob/issue-1-c3690fa5ac64/docs/COMPETITOR-RESEARCH.md) records the reproducible GitHub search method, ten representative open-source projects, capability comparison, and the design consequences adopted here.
- [`docs/ISSUE-1-DEVELOPMENT-NOTES.md`](https://github.com/konard/vietnam-accomodation-search/blob/issue-1-c3690fa5ac64/docs/ISSUE-1-DEVELOPMENT-NOTES.md) distills useful architecture, parser, security, Telegram-boundary, and validation findings from the development logs.

## Verification

- `npm run test:coverage` — 559 passed; every `src/**/*.js` module has 100% line coverage
- `bun test --timeout 30000` — 559 passed
- `deno test --allow-read` — 460 passed
- `npm run check` — lint, formatting, workflow checks, and duplication checks passed
- GitHub Actions now enforces the 100% product-line coverage gate on Ubuntu/Node in addition to the existing runtime/platform matrix

This is a bot/library/CLI change with no new visual UI, so before/after UI screenshots are not applicable.

## Access and security boundaries

Public Telegram history is collected through public web previews back to the rolling two-month cutoff. Telegram does not let bots fetch arbitrary private history; private communities must add the bot with appropriate permissions, and older private messages must be forwarded or imported. The optional MTProto user session is used only for an explicitly requested owner inquiry.

Official-property links must be explicitly marked or labeled. Before direct navigation, the collector rejects credentials, localhost/local names, private or link-local IPv4 literals, and loopback, mapped, link-local, unique-local, or multicast IPv6 literals. Consent pages, CAPTCHAs, access failures, source terms, and rate limits are respected rather than circumvented.

### CODE REVIEW

**Comment by konard** [comment]
## Working session summary

Implemented and finalized the accommodation search bot in [PR #2](https://github.com/konard/vietnam-accomodation-search/pull/2).

Key results:

- Browser Commander–based web and Telegram collection.
- Ranked 20+20 source registry with `/update_sources`.
- `/search --cheapest [N]`, VND conversion, photo limits, two-month Telegram filtering.
- Links Notation persistence and 10 GiB cache management.
- Fixed CodeQL’s unbounded price-regex finding.
- Fixed browser/desktop example builds by isolating server dependencies.
- Enabled repository dependency graph/security alerts.

Verification:

- Node: 493 tests passed
- Bun: 493 tests passed

<details>
<summary>Rest of the working session summary (238 characters)</summary>

```text
- Deno: 394 tests passed
- Formatting, lint, duplication, web build, CodeQL, dependency review, desktop builds, and all GitHub Actions workflows pass.
- Branch is current with `main`, working tree is clean, and the PR is ready for review.
```

</details>

---
*This summary was automatically extracted from the AI working session output.*

**Comment by konard** [comment]
## 🤖 Solution Draft Log
This log file contains the complete execution trace of the AI solution draft process.

### 💰 **Cost estimation:**
- Model: GPT-5.6 Sol
- Provider: OpenAI
- Public pricing estimate: $23.931718

### 📊 **Context and tokens usage:**
- 496.4K / 200K (248%) input tokens, 94.4K / 128K (74%) output tokens

Total: (496.4K + 21.4M cached) input tokens, 94.4K output tokens, $23.931718 cost

### 🤖 **Models used:**
- Tool: OpenAI Codex
- Requested: `gpt-5.6-sol`
- Thinking level: xhigh (~31999 tokens)
- **Model: GPT-5.6 Sol** (`gpt-5.6-sol`)

### 📎 **Log file uploaded as Gist** (8644KB)
- [View complete solution draft log](https://gist.githubusercontent.com/konard/774acefab63b3833bd7082c552eeec09/raw/22acad27efe423df0c0d24cddc48b348dfc6ecae/tmp-hive-mind-log-upload-O5ves4-sanitized.log.txt)

---
*Now working session is ended, feel free to review and add any feedback on the solution draft.*

**Comment by konard** [comment]
## ✅ Ready to merge

This pull request is now ready to be merged:
- All CI checks have passed
- No merge conflicts
- No pending changes

---
*Monitored by hive-mind with --auto-restart-until-mergeable flag*

**Comment by konard** [comment]
We also need to have additional 20 sources with focus on Nha Trang like:

@Arenda_Vetnam
@Arenda_Nyachang_Zhilye
@NaChangAp

With ability to search for similar, and all the messages for last 2-3 months should be fully parsable for all data they have, so it will be later possible to do search on any of parameters.

We also need to add a draft that uses user token for availability check. To contact the owner of accommodation via telegram private messages and ask if it is available.

Double check that requirements of https://github.com/konard/vietnam-accomodation-search/issues/1 are fully implemented in the maximum possible sense.

**Comment by konard** [comment]
🤖 **AI Work Session Started**

Starting automated work session at 2026-09-21T11:46:57.268Z

The PR has been converted to draft mode while work is in progress.

_This comment marks the beginning of an AI work session. Please wait for the session to finish, and provide your feedback._

_Runtime: solve `v2.29.0` · tool `codex` · model `gpt-5.6-sol` · task image `konard/hive-mind-dind:2.29.0@sha256:11a236d03854cbaea86fdd7ca06265c912cf0b67d1fcc9b3eb60bd84c43ff749`_

**Comment by konard** [comment]
## Working session summary

Implemented and finalized [PR #2](https://github.com/konard/vietnam-accomodation-search/pull/2).

Highlights:

- Added 20 additional Nha Trang Telegram sources, multilingual discovery, two-month pagination, and searchable listing parameters.
- Added explicit user-session availability inquiries through mtcute.
- Fixed broken documentation link and two CodeQL ReDoS findings with regression coverage.
- Local checks pass: Node 511, Bun 511, Deno 412, lint/format/duplication.
- All GitHub Actions workflows pass on `d1ad624`.
- Branch is current with `main`, working tree is clean, PR is mergeable and ready for review.

---
*This summary was automatically extracted from the AI working session output.*

**Comment by konard** [comment]
## 🤖 Solution Draft Log
This log file contains the complete execution trace of the AI solution draft process.

### 💰 **Cost estimation:**
- Model: GPT-5.6 Sol
- Provider: OpenAI
- Public pricing estimate: $18.307528

### 📊 **Context and tokens usage:**
- 425.5K / 200K (213%) input tokens, 67.6K / 128K (53%) output tokens

Total: (425.5K + 16.1M cached) input tokens, 67.6K output tokens, $18.307528 cost

### 🤖 **Models used:**
- Tool: OpenAI Codex
- Requested: `gpt-5.6-sol`
- Thinking level: xhigh (~31999 tokens)
- **Model: GPT-5.6 Sol** (`gpt-5.6-sol`)

### 📎 **Log file uploaded as Gist** (6061KB)
- [View complete solution draft log](https://gist.githubusercontent.com/konard/2ff1125ccac6bc33dc9e80745436519f/raw/8a453b5f6199264119872713c079c052cf0b75ed/tmp-hive-mind-log-upload-gjAbxy-sanitized.log.txt)

---
*Now working session is ended, feel free to review and add any feedback on the solution draft.*

**Comment by konard** [comment]
## ✅ Ready to merge

This pull request is now ready to be merged:
- All CI checks have passed
- No merge conflicts
- No pending changes

---
*Monitored by hive-mind with --auto-restart-until-mergeable flag*

**Comment by konard** [comment]
We need to make sure we support also deduplication, and merging and prices changes detection on official websites of each accommodation.

We also check for all open source competitors that provide same or similar features and make sure we support them all, also have list of competitors and everything useful from our past logs in comments, that might be useful with future development.

Make sure we have 100% test coverage and advanced parsing support.

**Comment by konard** [comment]
🤖 **AI Work Session Started**

Starting automated work session at 2026-09-21T12:54:12.775Z

The PR has been converted to draft mode while work is in progress.

_This comment marks the beginning of an AI work session. Please wait for the session to finish, and provide your feedback._

_Runtime: solve `v2.29.0` · tool `codex` · model `gpt-5.6-sol` · task image `konard/hive-mind-dind:2.29.0@sha256:11a236d03854cbaea86fdd7ca06265c912cf0b67d1fcc9b3eb60bd84c43ff749`_

**Comment by konard** [comment]
Implemented the requested follow-up in [`0aa02ab`](https://github.com/konard/vietnam-accomodation-search/commit/0aa02ab) with the durable research notes in [`5e9115d`](https://github.com/konard/vietnam-accomodation-search/commit/5e9115d).

Useful findings distilled from the development logs:

- Deduplication must be transitive and persist learned aliases; otherwise a later URL or source-ID observation can split a previously merged property again.
- A shared agent phone number is not sufficient identity. The conservative fallback also requires location plus a structural/title discriminator, while exact official/platform URLs and property IDs remain stronger keys.
- Marketplace and Telegram observations belong in price history, but only a page directly visited with `official-web` provenance can produce an official-property price-change event.
- Official URLs must be explicitly marked or labeled, remain separate from marketplace URLs, and pass local/private literal-address checks before browser navigation.
- The shared English/Vietnamese/Russian parser now retains both normalized fields and all unfamiliar labeled fields; its attacker-controlled spans are bounded following the earlier CodeQL feedback.

The focused reconciliation and official-site regressions were written red-first, including marketplace/official alias retention, official URL retention across later partial updates, unlabeled-URL rejection, unsafe IPv6 forms, and following an official link from an unpriced marketplace card.

Full local verification on the pushed SHA `5e9115d`: Node 558 tests with 100% product line coverage, Bun 558, Deno 459, and `npm run check` passing. The [competitor comparison](https://github.com/konard/vietnam-accomodation-search/blob/issue-1-c3690fa5ac64/docs/COMPETITOR-RESEARCH.md) and [development notes](https://github.com/konard/vietnam-accomodation-search/blob/issue-1-c3690fa5ac64/docs/ISSUE-1-DEVELOPMENT-NOTES.md) contain the reproducible research method and longer-lived conclusions.

File: src/listing-parser.js [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  const officialLabel =
+    /(?:official(?:\s+(?:website|site))?|property\s+(?:website|site)|website|trang\s*(?:web\s*)?chính\s*thức|официальн\p{L}*\s+сайт)\s*[:：-]/iu;
+  for (const line of text.split(/\r?\n/u)) {
+    const label = line.match(officialLabel);
+    if (!label) {
+      continue;
+    }
+    const value = line.slice((label.index || 0) + label[0].length);
+    for (const match of value.matchAll(/https?:\/\/[^\s<>()]+/giu)) {
+      const candidate = match[0].replace(/[.,;!?]+$/u, '');
```
[the remark this answers is not in the document]
Comment (konard) [reply src/listing-parser.js]: Fixed in a999250. I first added a 20,000-character adversarial regression: the old anchored punctuation regex took about 1.4 seconds and failed the 150 ms bound. The replacement performs one bounded scan, the focused case is about 28 ms locally, and the complete suite passes with 100% source line coverage.

**Comment by konard** [comment]
## 🤖 Solution Draft Log
This log file contains the complete execution trace of the AI solution draft process.

### 💰 **Cost estimation:**
- Model: GPT-5.6 Sol
- Provider: OpenAI
- Public pricing estimate: $27.833980

### 📊 **Context and tokens usage:**
- 1.2M / 200K (614%) input tokens, 114.8K / 128K (90%) output tokens

Total: (1.2M + 18.2M cached) input tokens, 114.8K output tokens, $27.833980 cost

### 🤖 **Models used:**
- Tool: OpenAI Codex
- Requested: `gpt-5.6-sol`
- Thinking level: xhigh (~31999 tokens)
- **Model: GPT-5.6 Sol** (`gpt-5.6-sol`)

### 📎 **Log file uploaded as Gist** (10031KB)
- [View complete solution draft log](https://gist.githubusercontent.com/konard/31142614d4fc07abdf4082a25fb42888/raw/663d290702f8ac777b6bd2ba857934850d06709e/tmp-hive-mind-log-upload-0VlUri-sanitized.log.txt)

---
*Now working session is ended, feel free to review and add any feedback on the solution draft.*

**Comment by konard** [comment]
## ✅ Ready to merge

This pull request is now ready to be merged:
- All CI checks have passed
- No merge conflicts
- No pending changes

---
*Monitored by hive-mind with --auto-restart-until-mergeable flag*
